src/_image_wrapper.cpp

```diff
@@ -224,50 +213,34 @@ image_resample(PyObject *self, PyObject* args, PyObject *kwargs)
         }
     }
 
-    if (ndim == 3) {
-        switch (PyArray_TYPE(input)) {
-        case NPY_UINT8:
-        case NPY_INT8:
-            resample<agg::rgba8>(input, output, params);
-            break;
-        case NPY_UINT16:
-        case NPY_INT16:
-            resample<agg::rgba16>(input, output, params);
-            break;
-        case NPY_FLOAT32:
-            resample<agg::rgba32>(input, output, params);
-            break;
-        case NPY_FLOAT64:
-            resample<agg::rgba64>(input, output, params);
-            break;
-        default:
-            PyErr_SetString(
-                PyExc_ValueError,
-                "arrays must be of dtype byte, short, float32 or float64");
-            goto error;
-        }
-    } else { // ndim == 2
-        switch (PyArray_TYPE(input)) {
-        case NPY_UINT8:
-        case NPY_INT8:
-            resample<unsigned char>(input, output, params);
-            break;
-        case NPY_UINT16:
-        case NPY_INT16:
-            resample<unsigned short>(input, output, params);
-            break;
-        case NPY_FLOAT32:
-            resample<float>(input, output, params);
-            break;
-        case NPY_FLOAT64:
-            resample<double>(input, output, params);
-            break;
-        default:
-            PyErr_SetString(
-                PyExc_ValueError,
-                "arrays must be of dtype byte, short, float32 or float64");
-            goto error;
-        }
+    if (auto resampler =
+            (ndim == 2) ? (
+                (type == NPY_UINT8) ? resample<agg::gray8> :
+                (type == NPY_INT8) ? resample<agg::gray8> :
+                (type == NPY_UINT16) ? resample<agg::gray16> :
+                (type == NPY_INT16) ? resample<agg::gray16> :
+                (type == NPY_FLOAT32) ? resample<agg::gray32> :
+                (type == NPY_FLOAT64) ? resample<agg::gray64> :
+                nullptr) : (
+            // ndim == 3
+                (type == NPY_UINT8) ? resample<agg::rgba8> :
+                (type == NPY_INT8) ? resample<agg::rgba8> :
+                (type == NPY_UINT16) ? resample<agg::rgba16> :
+                (type == NPY_INT16) ? resample<agg::rgba16> :
+                (type == NPY_FLOAT32) ? resample<agg::rgba32> :
+                (type == NPY_FLOAT64) ? resample<agg::rgba64> :
+                nullptr)) {
```

Comment:
Is there an obvious benefit to write it is this way?

(In addition to "hide" that there are no tests for short? Which I realize is not the purpose, but a direct consequence.  A single raise I can see, but other than that, the previous code was easier to read for me anyway, But then I know that you like these constructs in general (and I don't...), so I can ignore that if there is a benefit.)

Reply:
For me the point of writing it like this is to allow focusing on what matters (the NPY_FOO -> agg:type mapping, each pair of which now appears on a single line), and reduce line noise (input/output/params, which are the same every time, and the breaks; which add no info).

The fact that this "hides" that a branch is not taken is not a plus (if anything it's a minus), we should enable branch coverage in C++ too if possible.